File: 57. Insert Interval.exs
defmodule Solution do
  def insert(intervals, new_interval) do
    [new_interval | intervals]
    |> Enum.sort()
    |> merge([])
  end

  def merge([], result), do: Enum.reverse(result)
  def merge([[s, e] | tail], []), do: merge(tail, [[s, e]])
  def merge([[s, e] | tail], [[ps, pe] | p_tail] = rest) do
    if s <= pe do
      merge(tail, [[ps, max(pe, e)] | p_tail])
    else
      merge(tail, [[s, e] | rest])
    end
  end
end

File: 57. Insert Interval.rb
#*57. Insert Interval
# @param {Integer[][]} intervals
# @param {Integer[]} new_interval
# @return {Integer[][]}
#*SOLUTION
def insert(intervals = [[1,3],[6,9]], new_interval = [2,5])
  intervals << new_interval
  intervals.sort!
  
  res = [intervals[0]]
  l, r = res[0]
  
  (intervals.size - 1).times do |i|
    if r >= intervals[i+1][0]
      r = [r, intervals[i+1][1]].max
      res[-1][1] = r
    else
      l, r = intervals[i+1]
      res << [l, r]
    end
  end
  res
end
p insert